ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/DefaultBlockchain.java

```diff
@@ -320,58 +326,82 @@ private BlockAddedEvent handleFork(final BlockchainStorage.Updater updater, fina
   }
 
   private BlockAddedEvent handleChainReorg(
-      final BlockchainStorage.Updater updater, final Block newChainHead) {
-    final Hash oldChainHead = blockchainStorage.getChainHead().get();
-    BlockHeader oldChain = blockchainStorage.getBlockHeader(oldChainHead).get();
-    BlockHeader newChain = newChainHead.getHeader();
+      final BlockchainStorage.Updater updater, final BlockWithReceipts newChainHeadWithReceipts) {
+    final Block newChainHeadBlock = newChainHeadWithReceipts.getBlock();
+    final Block oldChainHeadBlock = getChainHeadBlock();
+    BlockHeader currentOldChainHeader = oldChainHeadBlock.getHeader();
+    BlockHeader currentNewChainHeader = newChainHeadBlock.getHeader();
 
     // Update chain head
-    updater.setChainHead(newChain.getHash());
+    updater.setChainHead(newChainHeadBlock.getHash());
 
-    // Track transactions to be added and removed
+    // Track transactions and logs to be added and removed
     final Map<Hash, List<Transaction>> newTransactions = new HashMap<>();
     final List<Transaction> removedTransactions = new ArrayList<>();
+    final Deque<LogWithMetadata> logsWithMetadata = new ArrayDeque<>();
 
-    while (newChain.getNumber() > oldChain.getNumber()) {
+    while (currentNewChainHeader.getNumber() > currentOldChainHeader.getNumber()) {
       // If new chain is longer than old chain, walk back until we meet the old chain by number
       // adding indexing for new chain along the way.
-      final Hash blockHash = newChain.getHash();
-      updater.putBlockHash(newChain.getNumber(), blockHash);
+      final Hash blockHash = currentNewChainHeader.getHash();
+      updater.putBlockHash(currentNewChainHeader.getNumber(), blockHash);
       final List<Transaction> newTxs =
-          blockHash.equals(newChainHead.getHash())
-              ? newChainHead.getBody().getTransactions()
-              : blockchainStorage.getBlockBody(blockHash).get().getTransactions();
+          blockchainStorage
+              .getBlockBody(blockHash)
+              .map(BlockBody::getTransactions)
+              // If it's not in storage, it must be the new chain head
+              .orElseGet(() -> newChainHeadBlock.getBody().getTransactions());
       newTransactions.put(blockHash, newTxs);
 
-      newChain = blockchainStorage.getBlockHeader(newChain.getParentHash()).get();
+      addAddedLogsWithMetadata(
+          logsWithMetadata,
+          getBlockWithReceipts(currentNewChainHeader).orElse(newChainHeadWithReceipts));
```

Comment:
We're querying for the block body twice here.  We should just execute one query to get the `BlockWithReceipts`, and pull the transactions from there.  Maybe this could be simplified by keeping the old chain / new chain pointer variables (`currentOldChainHeader` and `currentNewChainHeader`) as `BlockWithReceipts`'s.